### PR TITLE
Add GSC

### PR DESCRIPTION
**Pull request by Jake-NotTheMuss**
## Description

GSC is a programming language used for Call of Duty's virtual machine.
## Checklist:

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      
      - .gsc extension: https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Agsc+NOT+nothack
      - .csc extension https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Acsc+NOT+nothack
      - .gsh extension: https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Agsh+NOT+nothack
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - samples:
      - (.csc) https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/samples/zm_init.gsc
      - (.gsc) https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/samples/math_shared.gsc
      - (.csc) https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/samples/_spiderbot_ride.csc
      - (.gsc) https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/samples/struct.gsc
      - (.gsh) https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/samples/array_override_common.gsh
      - (.gsc) https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/samples/array_override_common.gsc
      - License for these sources: https://github.com/Jake-NotTheMuss/CoDT7-Sublime/blob/master/LICENSE 
  
  - [x] I have included a syntax highlighting grammar: https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=xml&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2FJake-NotTheMuss%2FCoDT7-Sublime%2Fmaster%2Fother%2Fgsc.tmLanguage&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fraw.githubusercontent.com%2FJoshr520%2Fee_patch%2Fmain%2Fscripts%2Fzm%2Farray_override%2Farray_override_common.gsc&code=
  - [x] I have updated the heuristics to distinguish my language from others using the same extension.

### CODE REVIEW

**Review by lildude — CHANGES_REQUESTED** [review]
You've only listed three sample files in the OP but have added more. Please update the OP to list where you've obtain _all_ of the sample files and their respective licenses.

**Comment by Jake-NotTheMuss** [comment]
The other files are directly from the Black Ops 3 mod tools. I think it's GNU v3 but I'll install Black Ops 3 to make sure and update with the license for those files when I confirm.

**Comment by Jake-NotTheMuss** [comment]
I've added the license for the other samples, but the syntax highlighting grammar that was given is provided in the mod tools, so I guess that means it's licensed under GPL 3, so I will create my own with an acceptable license.

**Comment by lildude** [comment]
> I've added the license for the other samples, but the syntax highlighting grammar that was given is provided in the mod tools, so I guess that means it's licensed under GPL 3, so I will create my own with an acceptable license.

Thanks. Can you update the OP to point to the source of the files too, as you've done with the current three files.

As for the grammar, the one you've added - https://github.com/Jake-NotTheMuss/CoDT7-Sublime.git - is covered by the Unlicense License, which is fine.

I still need to check usage levels of all extensions meet our requirements.

**Comment by Jake-NotTheMuss** [comment]
Those scripts aren't on the Internet. Should I create a repo and push those files to it? And the license of the sample files can be anything, right?

**Comment by Jake-NotTheMuss** [comment]
The rest of the sample files and their license have been added.

**Comment by Jake-NotTheMuss** [comment]
Should I add tests for each GSC extension to test/test_heuristics.rb?

**Comment by Jake-NotTheMuss** [comment]
@lildude

**Comment by lildude** [comment]
> Should I add tests for each GSC extension to test/test_heuristics.rb?

That's what the test failure is telling you 😉 

Or you can remove the heuristics if these extensions are unique to this language.

**Comment by Jake-NotTheMuss** [comment]
@lildude I added the heuristic tests for each extension. These extensions are not associated with any other supported language on github, but they are there just be certain non-GSC .gsc files aren't counted.

**Comment by Jake-NotTheMuss** [comment]
@lildude Do I need to run the tests again or is there something else I need to do first?

**Comment by Jake-NotTheMuss** [comment]
@Alhadis

File: samples/GSC/array_shared.gsc [hunk 1]
```diff
@@ -0,0 +1,1321 @@
+#using scripts\codescripts\struct;
```
Comment (lildude) [inline samples/GSC/array_shared.gsc:1]: This file is massive and I don't think it adds much more than the sample above. Please remove it.

File: samples/GSC/scene_shared.csc [hunk 2]
```diff
@@ -0,0 +1,2390 @@
+#using scripts\codescripts\struct;
```
Comment (lildude) [inline samples/GSC/scene_shared.csc:1]: Wow!! This file is huge and seems to nearly repeat similar segments of code. Are you able to replace this with a smaller representative sample with less apparent duplication?

File: samples/GSC/shared.gsh [hunk 3]
```diff
@@ -0,0 +1,935 @@
+#define PLAYER_1 0
```
Comment (lildude) [inline samples/GSC/shared.gsh:1]: Are you able to find a smaller sample? This file is mostly filled with `define` lines which don't differ by very much syntactically.

**Review by lildude — CHANGES_REQUESTED** [review]
This PR is probably close to being merged based on the new temp limits detailed in https://github.com/github/linguist/issues/5756 however these samples are quite massive.

**Comment by Jake-NotTheMuss** [comment]
> This PR is probably close to being merged based on the new temp limits detailed in #5756 however these samples are quite massive.

Samples should be a good size now

**Comment by lildude** [comment]
@Alhadis would you mind taking a quick look at the regexes in this PR when you've got a mo? 🙇

**Review by Alhadis — CHANGES_REQUESTED** [review]

**Comment by Jake-NotTheMuss** [comment]
@lildude @Alhadis

**Review by Alhadis — APPROVED** [review]

**Pull request merged**